lib/lf/uscalfe/fe_space_hp.h

```diff
@@ -0,0 +1,128 @@
+ */
+template <typename SCALAR>
+class FeSpaceHP : public ScalarFESpace<SCALAR> {
+ public:
+  using Scalar = SCALAR;
+
+  FeSpaceHP() = delete;
+  FeSpaceHP(const FeSpaceHP &) = delete;
+  FeSpaceHP(FeSpaceHP &&) noexcept = default;
+  FeSpaceHP &operator=(const FeSpaceHP &) = delete;
+  FeSpaceHP &operator=(FeSpaceHP &&) noexcept = default;
+
+  /**
+   * @brief Constructor: Sets up the dof handler
+   * @param mesh_p A shared pointer to the underlying mesh (immutable)
+   * @param N The polynomial degree of the Finite Element Space
+   */
+  explicit FeSpaceHP(std::shared_ptr<const lf::mesh::Mesh> mesh_p, unsigned N)
+      : ScalarFESpace<SCALAR>(std::move(mesh_p)),
+        ref_el_(mesh_p, nullptr),
+        degree_(N) {
+    init();
+  }
+
+  using ScalarFESpace<SCALAR>::Mesh;
+
+  /** @brief access to associated local-to-global map
+   * @return a reference to the lf::assemble::DofHandler object (immutable)
+   */
+  [[nodiscard]] const lf::assemble::DofHandler &LocGlobMap() const override {
+    LF_VERIFY_MSG(Mesh() != nullptr, "No valid FE space object: no mesh");
+    LF_VERIFY_MSG(dofh_p_ != nullptr,
+                  "No valid FE space object: no dof handler");
+    return *dofh_p_;
+  }
+
+  /** @brief access to shape function layout for cells
+   * @copydoc SclarFESpace::ShapeFunctionLayout(const lf::mesh::Entity&)
+   */
+  [[nodiscard]] std::shared_ptr<const ScalarReferenceFiniteElement<SCALAR>>
+  ShapeFunctionLayout(const lf::mesh::Entity &entity) const override {
+    return ref_el_(entity);
+  }
+
+  /** @brief number of _interior_ shape functions associated to entities of
+   * various types
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      const lf::mesh::Entity &entity) const override {
+    return ShapeFunctionLayout(entity)->NumRefShapeFunctions();
+  }
+
+  ~FeSpaceHP() override = default;
+
+ private:
+  // R.H.: Should this really be shared pointers ?
+  lf::mesh::utils::AllCodimMeshDataSet<
+      std::shared_ptr<const lf::uscalfe::ScalarReferenceFiniteElement<SCALAR>>>
+      ref_el_;
+  std::unique_ptr<lf::assemble::UniformFEDofHandler> dofh_p_;
+  unsigned degree_;
+
+  // R.H. Detailed comment needed for this function
+  void init() {
+    // Initialize all shape function layouts for nodes
+    size_type num_rsf_node = 1;
+    for (auto entity : Mesh()->Entities(2)) {
+      ref_el_(*entity) = std::make_shared<FeLagrangePoint<SCALAR>>(degree_);
+    }
+    // Initialize all shape function layouts for the edges
+    size_type num_rsf_edge = 0;
+    for (auto entity : Mesh()->Entities(1)) {
+      ref_el_(*entity) = std::make_shared<FeHPSegment<SCALAR>>(
+          degree_, entity->RelativeOrientations());
+      num_rsf_edge = ref_el_(*entity)->NumRefShapeFunctions(0);
+    }
+    // Initialize all shape function layouts for the cells
+    size_type num_rsf_tria = 0;
+    size_type num_rsf_quad = 0;
+    for (auto entity : Mesh()->Entities(0)) {
```

Comment:
This is a dangerous "hybrid design". On theone hand, a uniform distributions of degrees of freedom is assumed by relying on `f::assemble::UniformFEDofHandler`, but on the other hand every cells gets its own reference element. 

Would it be possible to preallocate a small number of reference elements and then give every entity a pointer to the right one.

Reply:
That's exactly one of the reasons why it's hard to implement hp-fem ;)

In principle it is possible to rely on a small number of reference finite elements in order to solve the orientation problem. However neither NgSolve, HyDi nor Deal.II have relied on this technique because it is quite cumbersome to implement (especially in 3d) and doesn't make much sense if you want to vary the polynomial degree inside the mesh.
For uniform polynomial degree , you would need to implement `2^3 = 8` different reference elements for triangle and 16 for quadrilateral. With the help of affine mappings you can reduce this number a bit but the implementation gets even more "cumbersome"...

I wouldn't recommend to go this route because it makes it nearly impossible to extend the implementation at a later time to vary the polynomial degree over the mesh. Even though the implementation would be slightly faster this way.